### PR TITLE
[FEATURE] Global Datasource CRUD + DRY refactoring

### PR DESCRIPTION
**Pull request by celian-garcia**
Refactoring on top of the global datasource CRUD PR #1319  to use generics like we do for the global/project variables.

image
Ref #1033 
Closes #1319 (Let's see if we can automatically close a PR but I'm not sure)
# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [x] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [x] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [x] Visual tests are stable and unlikely to be flaky. See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests) and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

### CODE REVIEW

File: ui/app/src/components/DatasourceList/DatasourceDrawer.tsx [hunk 1]
```diff
@@ -21,16 +21,19 @@ import { DatasourceEditorForm } from './DatasourceEditorForm';
 
 export type ActionStr = 'Create' | 'Update';
 
-interface DatasourceDrawerProps {
-  datasource: Datasource;
+type DispatchWithPromise<A> = (value: A) => Promise<void>;
```
Comment (Gladorme) [inline ui/app/src/components/DatasourceList/DatasourceDrawer.tsx:24]: Move to a utils file?
I remember seeing this same type declared in the variable PR :p

File: ui/app/src/views/projects/ProjectTabs.tsx [hunk 2]
```diff
@@ -248,13 +248,13 @@ export function ProjectTabs(props: DashboardVariableTabsProps) {
         </Tabs>
         <TabButton index={value} projectName={projectName} />
       </Stack>
-      <TabPanel value={value} index="dashboards">
+      <TabPanel value={value} index={dashboardTabIndex}>
```
Comment (Gladorme) [inline ui/app/src/views/projects/ProjectTabs.tsx:251]: dashboard**s**TabIndex

**Review by sjcobb — APPROVED** [review]
💯

File: ui/app/src/components/DatasourceList/DatasourceEditorForm.tsx [hunk 3]
```diff
@@ -60,20 +63,21 @@ function getInitialState(datasource: Datasource): Datasource {
...
   onClose: DispatchWithoutAction;
-  onDelete: DispatchWithoutAction | undefined;
+  onDelete?: DispatchWithoutAction;
 }
 
-export function DatasourceEditorForm(props: DatasourceEditorFormProps) {
+export function DatasourceEditorForm<T extends Datasource>(props: DatasourceEditorFormProps<T>) {
   const { initialDatasource, saveActionStr, onSave, onClose, onDelete } = props;
+
+  const patchedInitialDatasource = getPatchedInitialProjectDs(initialDatasource);
```
[the remark this answers is not in the document]
Comment (celian-garcia) [reply ui/app/src/components/DatasourceList/DatasourceEditorForm.tsx:77]: I´d agree. Actually the ``getInitialState`` was not bad either. I can also remove Project. But definitely shortcuts can bring confusion

[the remark this answers is not in the document]
Comment (celian-garcia) [reply ui/app/src/components/DatasourceList/DatasourceEditorForm.tsx]: Renamed to ``getInitialState`` to be the same as in Variables CRUD

File: ui/core/src/utils/text.ts [hunk 4]
```diff
@@ -34,9 +34,9 @@ export function getVariableDisplayName(variable: Variable) {
 /**
  * If the variable has a display name, return the datasource display name
  * Else, only return the datasource name
- * @param variable Project or Global datasource
+ * @param datasource Project or Global datasource
  */
-export function getDatasourceDisplayName(datasource: Datasource) {
+export function getDatasourceDisplayName(datasource: ProjectDatasource | GlobalDatasource) {
```
Comment (AntoineThebaud) [inline ui/core/src/utils/text.ts:39]: You should reuse the `Datasource` type you created here

`export type Datasource = ProjectDatasource | GlobalDatasource;`

File: ui/core/src/utils/text.ts [hunk 5]
```diff
@@ -67,9 +67,9 @@ export function getVariableExtendedDisplayName(variable: Variable) {
 /**
  * If the datasource has a display name, return the datasource display name and the datasource name
  * Else, only return the datasource name
- * @param variable Project or Global datasource
+ * @param datasource Project or Global datasource
  */
-export function getDatasourceExtendedDisplayName(datasource: Datasource) {
+export function getDatasourceExtendedDisplayName(datasource: ProjectDatasource | GlobalDatasource) {
```
Comment (AntoineThebaud) [inline ui/core/src/utils/text.ts:72]: same as above

File: ui/app/src/model/datasource-api.ts [hunk 6]
```diff
@@ -216,7 +216,7 @@ export class CachedDatasourceAPI implements DatasourceApi {
 }
 
 // Helper function for getting a proxy URL from a datasource or global datasource
-function getProxyUrl(datasource: Datasource | GlobalDatasource) {
+function getProxyUrl(datasource: ProjectDatasource | GlobalDatasource) {
```
Comment (AntoineThebaud) [inline ui/app/src/model/datasource-api.ts:219]: You should reuse the `Datasource` type you created here

`export type Datasource = ProjectDatasource | GlobalDatasource;`

**Comment by AntoineThebaud** [comment]
Just few minor remarks, otherwise looks good! Thanks for this awesome refactoring 🙌

File: ui/app/src/components/DatasourceList/DatasourceList.tsx [hunk 7]
```diff
@@ -71,70 +65,32 @@ export function DatasourceList(props: DatasourceListProperties) {
...
+  const [targetedDatasource, setTargetedDatasource] = useState<T>();
   const [isEditDatasourceFormStateOpened, setEditDatasourceFormStateOpened] = useState<boolean>(false);
 
   const handleDatasourceUpdate = useCallback(
-    (datasource: Datasource) => {
+    async (datasource: T) => {
       if (targetedDatasource != undefined && datasource.metadata.name != targetedDatasource.metadata.name) {
         // In this case "move" the datasource, aka create it with the new name & remove the former one
         // We do this because we can't just do a PUT call in that case (results in "document not found" error)
-        // TODO: create + delete calls should be bundled together so that we avoid the case where only one would succeed
```
[the remark this answers is not in the document]
Comment (celian-garcia) [reply ui/app/src/components/DatasourceList/DatasourceList.tsx:84]: :astonished: Absolutely ! I didn't notice it thanks

**Comment by celian-garcia** [comment]
> Just few minor remarks, otherwise looks good! Thanks for this awesome refactoring 🙌 This will help a lot with the dashboard datasources CRUD implementation in the future.

Thanks to you for the review @AntoineThebaud :heart: You may have noticed that I finally made the choice to put on top of all the components the backend queries. This was an option we discussed already and we were not sure. Now seeing it live, I think we won in readability :)

**Pull request merged**